src/filters/array.rs

```diff
@@ -0,0 +1,171 @@
+use liquid_value::Value;
+
+use super::{check_args_len, invalid_input};
+use interpreter::FilterResult;
+
+/// Receives a `Value::Array` as an input.
+/// Returns a copy of the input with the given value appended at the end.
+pub fn push(input: &Value, args: &[Value]) -> FilterResult {
+    check_args_len(args, 1, 0)?;
+
+    let value = args[0].clone();
+    let mut array = input
+        .as_array()
+        .ok_or_else(|| invalid_input("Array expected"))?
+        .clone();
+    array.push(value);
+
+    Ok(Value::array(array))
+}
+
+/// Receives a `Value::Array` as an input.
+/// Returns a copy of the input with the last element removed.
+pub fn pop(input: &Value, args: &[Value]) -> FilterResult {
+    check_args_len(args, 0, 0)?;
+
+    let mut array = input
+        .as_array()
+        .ok_or_else(|| invalid_input("Array expected"))?
+        .clone();
+    array.pop();
+
+    Ok(Value::array(array))
+}
+
+/// Receives a `Value::Array` as an input.
+/// Returns a copy of the input with the given value appended at the beginning.
+pub fn unshift(input: &Value, args: &[Value]) -> FilterResult {
+    check_args_len(args, 1, 0)?;
+
+    let value = args[0].clone();
+    let mut array = input
+        .as_array()
+        .ok_or_else(|| invalid_input("Array expected"))?
+        .clone();
+    array.insert(0, value);
+
+    Ok(Value::array(array))
+}
+
+/// Receives a `Value::Array` as an input.
+/// Returns a copy of the input with the first element removed.
+pub fn shift(input: &Value, args: &[Value]) -> FilterResult {
+    check_args_len(args, 0, 0)?;
+
+    let mut array = input
+        .as_array()
+        .ok_or_else(|| invalid_input("Array expected"))?
+        .clone();
+    array.remove(0);
+
+    Ok(Value::array(array))
+}
+
+/// Convert an array into a sentence.
+pub fn array_to_sentence_string(input: &Value, args: &[Value]) -> FilterResult {
+    check_args_len(args, 0, 1)?;
+
+    let connector = args
+        .get(0)
+        .and_then(|arg| Some(arg.to_string()))
+        .unwrap_or_else(|| "and".to_string());
+
+    let mut array = input
+        .as_array()
+        .ok_or_else(|| invalid_input("Array expected"))?
+        .iter();
+
+    let len = array.len();
+    match len {
+        0 => Ok(Value::scalar("")),
+        1 => Ok(Value::scalar(array.next().unwrap().to_string())),
+        _ => {
+            let first = array.next().unwrap().to_string();
+            let last = array.next_back().unwrap().to_string();
```

Comment:
Wouldn't `to_str` be preferred since we don't an owned version

And granted, while performance for this isn't as critical, you could take it a step further and `write!` into a buffer for this entire operation.  That'll remove all of the allocations from `to_string` and `format` (O(N) allocations) and instead have the allocations of the `Vec` growing which is usually optimized pretty well.

Doing this, you *might* even be able to merge the code paths or 0, 1, and n size arrays.